sockeye/data_io.py

```diff
@@ -484,13 +501,12 @@ def prepare_data(source: str, target: str,
                  output_prefix: str,
                  keep_tmp_shard_files: bool = False):
     logger.info("Preparing data.")
-
-    # write vocabularies
-    vocab.vocab_to_json(vocab_source, os.path.join(output_prefix, C.VOCAB_SRC_NAME) + C.JSON_SUFFIX)
-    vocab.vocab_to_json(vocab_target, os.path.join(output_prefix, C.VOCAB_TRG_NAME) + C.JSON_SUFFIX)
+    # write vocabularies to data folder
+    vocab.save_source_vocabs(source_vocabs, output_prefix)
+    vocab.vocab_to_json(target_vocab, os.path.join(output_prefix, C.VOCAB_TRG_NAME))
 
     # Pass 1: get target/source length ratios.
-    length_statistics = analyze_sequence_lengths(source, target, vocab_source, vocab_target,
+    length_statistics = analyze_sequence_lengths(source_fnames[0], target_fname, source_vocabs[0], target_vocab,
```

Comment:
`analyze_sequence_lengths ` also checks that the files are parallel. So here we do not actually check that the source factors have as many sentences as the other files. If the source factors contain fewer elements than the other files the shard_data below will not consume all training sentences, rather than complain, no? I think it would be good to check the length somewhere. Maybe it would make sense to do that check at the end of `shard_data` (by checking that all SentenceReaders have finished).

Reply:
Actually, would it make sense then to just modify analyze_sequence_lengths to take `sourceS` like all the other functions? We can then check for token parallelism and line parallelism (by checking if all readers are done). This covers all use cases (prepare data, training iter, val iter).

Reply:
sounds good to me.